beamer/agent/chain.py

```diff
@@ -572,6 +575,61 @@ def _timestamp_is_l1_finalized(
     return int(time.time()) > timestamp + target_chain_finality
 
 
+def maybe_prove(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    # mainnet: 10, goerli: 420, local: 17
+    if context.target_chain_id not in [10, 420, 17]:
+        return False
+
+    own_challenge_stake = Wei(claim.get_challenger_stake(context.address))
+    can_prove = (
+        claim.claimer == context.address and claim.challenger_exists()
+    ) or own_challenge_stake > 0
+
+    if not can_prove:
+        return False
+
+    if claim.message_proved:
+        return False
+
+    if request.fill_tx:
+        proof_tx = request.fill_tx
+    elif claim.invalidation_tx:
+        proof_tx = claim.invalidation_tx
+
+    if proof_tx in context.l1_resolutions:
+        return False
+
+    future = context.task_pool.submit(
+        run_relayer_for_tx,
+        context.config.base_chain_rpc_url,
+        context.target_rpc_url,
+        context.source_rpc_url,
+        context.config.account.key,
+        proof_tx,
+        True,
+    )
+
+    def on_future_done(f: Future) -> None:
+        try:
+            f.result()
+            claim.message_proved = True
```

Comment:
Nitpick: I would move the
```python
            claim.message_proved = True
```
to an `else:` block of this `try:` block, to minimise the amount of code under `try`.

Reply:
What if the `relayer prove` process fails?
We will see the failure, log it, but the details will be missing, right?
Perhaps that is best we can do right now, but I'm wondering about situations where one sees `Optimism prove failed` error in agent logs and then... what? What would be the next steps?

Also, in what cases can the `relayer prove` process fail? (Apart from the obvious, e.g. lack of funds.)

Reply:
@GabrielBuragev can you answer this?

Reply:
> Also, in what cases can the `relayer prove` process fail? (Apart from the obvious, e.g. lack of funds.)

Excluding the failures related to incorrect arguments passed to the `relayer prove`, buggy rpcs, etc. (externalities) then this command can only fail if a transaction reverted for any possible reasons.

Reply:
> What if the relayer prove process fails?

From an agent's perspective, it only cares if the relayer did its job or not. If it failed for whatever reason, it will call the relayer again. See my comment below regarding better logging or information passing.